src/opensynth/models/faraday/gaussian_mixture/gmm_lightning.py

```diff
@@ -121,9 +123,8 @@ def training_step(self, batch: torch.Tensor) -> None:
             )
         responsibilities = log_responsibilities.exp()
 
-        responsibilities += 1e-15 / len(
-            encoded_batch
-        )  # Avoid zero responsibilities
+        # ensure the lowest cluster probability is 1/N.
+        responsibilities += 1 / (self.num_datapoints * len(encoded_batch))
```

Comment:
I think we want this to be 1 divided by the number of datapoints multiplied by the number of batches. This is because we want the minimum probability to be 1/number of datapoints, but then covariances get updated every training step (and there is 1 training step per batch), so we need to also divide by the number of batches (not the length of the batch). 

`1 / (self.num_datapoints * num_batches)` = `1 / (self.num_datapoints * self.num_datapoints / len(encoded_batch))` = `len(encoded_batch) / (self.num_datapoints**2)`

Reply:
Good point, you're right I have missed the multiplier for number of batches. I think there's still something missing though. 

What we want is for the minimum `component_weight` to be equal to `1/num_datapoints`. The component weight is equal to the responsibilities summed over the 1st dimension. The size of `responsibilities here is `len(encoded_batch), num_components`.  So if we're adding a constant `a` to each entry of the responsibilities, when computing the component weight we'll end up with `len(encoded_batch) * a` as the minimum probability. So I think we also need to divide by the length of the batch. So the final thing would be:

`1 / (self.num_datapoints * num_batches * len(encoded_batch)) == 1/self.num_datapoints**2` 

Does that make sense?

Reply:
Yes this makes sense, I get you! I agree with your logic :)